libdevice/fallback-gsort.cpp

```diff
@@ -0,0 +1,499 @@
+void __devicelib_default_work_group_joint_sort_ascending_p3u64_u32_p3i8(
+    uint64_t *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::less<uint64_t>{});
+}
+
+DEVICE_EXTERN_C_INLINE
+void __devicelib_default_work_group_joint_sort_descending_p1u64_u32_p1i8(
+    uint64_t *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::greater<uint64_t>{});
+}
+
+DEVICE_EXTERN_C_INLINE
+void __devicelib_default_work_group_joint_sort_descending_p1u64_u32_p3i8(
+    uint64_t *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::greater<uint64_t>{});
+}
+
+DEVICE_EXTERN_C_INLINE
+void __devicelib_default_work_group_joint_sort_descending_p3u64_u32_p1i8(
+    uint64_t *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::greater<uint64_t>{});
+}
+
+DEVICE_EXTERN_C_INLINE
+void __devicelib_default_work_group_joint_sort_descending_p3u64_u32_p3i8(
+    uint64_t *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::greater<uint64_t>{});
+}
+//=============== default work grop joint sort for fp32 ======================
+DEVICE_EXTERN_C_INLINE
+void __devicelib_default_work_group_joint_sort_ascending_p1f32_u32_p1i8(
+    float *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::less<float>{});
+}
+
+DEVICE_EXTERN_C_INLINE
+void __devicelib_default_work_group_joint_sort_ascending_p1f32_u32_p3i8(
+    float *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::less<float>{});
+}
+
+DEVICE_EXTERN_C_INLINE
+void __devicelib_default_work_group_joint_sort_ascending_p3f32_u32_p1i8(
+    float *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::less<float>{});
+}
+
+DEVICE_EXTERN_C_INLINE
+void __devicelib_default_work_group_joint_sort_ascending_p3f32_u32_p3i8(
+    float *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::less<float>{});
+}
+
+DEVICE_EXTERN_C_INLINE
+void __devicelib_default_work_group_joint_sort_descending_p1f32_u32_p1i8(
+    float *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::greater<float>{});
+}
+
+DEVICE_EXTERN_C_INLINE
+void __devicelib_default_work_group_joint_sort_descending_p1f32_u32_p3i8(
+    float *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::greater<float>{});
+}
+
+DEVICE_EXTERN_C_INLINE
+void __devicelib_default_work_group_joint_sort_descending_p3f32_u32_p1i8(
+    float *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::greater<float>{});
+}
+
+DEVICE_EXTERN_C_INLINE
+void __devicelib_default_work_group_joint_sort_descending_p3f32_u32_p3i8(
+    float *first, uint32_t n, uint8_t *scratch) {
+  merge_sort(first, n, scratch, std::greater<float>{});
+}
+
+//============ default work grop private sort for signed integer ==============
+// Since 'first' should point to 'private' memory address space, it can only be
```

Comment:
Hi, @andreyfe1 
For private sort, we only need to use 'p1' to decorate memory address space for 'first', no 'p3' version, is it correct?
Thanks very much.